src/biz/workspace/publish.rs

```diff
@@ -154,3 +120,296 @@ async fn check_workspace_namespace(new_namespace: &str) -> Result<(), AppError>
+    let metadata =
+      select_publish_collab_meta(&self.pg_pool, publish_namespace, publish_name).await?;
+    Ok(metadata)
+  }
+
+  async fn get_collab_with_view_metadata_by_view_id(
+    &self,
+    view_id: &Uuid,
+  ) -> Result<Option<(PublishViewMetaData, Vec<u8>)>, AppError> {
+    let result = match select_published_data_for_view_id(&self.pg_pool, view_id).await? {
+      Some((js_val, blob)) => {
+        let metadata = serde_json::from_value(js_val)?;
+        Ok(Some((metadata, blob)))
+      },
+      None => Ok(None),
+    };
+    if result.is_err() {
+      self.metrics.incr_failure_read_count(1);
+    } else {
+      self.metrics.incr_success_read_count(1);
+    }
+    result
+  }
+
+  async fn get_collab_publish_info(&self, view_id: &Uuid) -> Result<PublishInfo, AppError> {
+    select_published_collab_info(&self.pg_pool, view_id).await
+  }
+
+  async fn get_collab_blob_by_publish_namespace(
+    &self,
+    publish_namespace: &str,
+    publish_name: &str,
+  ) -> Result<Vec<u8>, AppError> {
+    let result = select_published_collab_blob(&self.pg_pool, publish_namespace, publish_name).await;
+    if result.is_err() {
+      self.metrics.incr_failure_read_count(1);
+    } else {
+      self.metrics.incr_success_read_count(1);
+    }
+    result
+  }
+
+  async fn delete_collab(
+    &self,
+    workspace_id: &Uuid,
+    view_ids: &[Uuid],
+    user_uuid: &Uuid,
+  ) -> Result<(), AppError> {
+    check_workspace_owner_or_publisher(&self.pg_pool, user_uuid, workspace_id, view_ids).await?;
+    delete_published_collabs(&self.pg_pool, workspace_id, view_ids).await?;
+    Ok(())
+  }
+}
+
+pub struct PublishedCollabS3StoreWithPostgresFallback {
+  metrics: Arc<PublishedCollabMetrics>,
+  pg_pool: PgPool,
+  bucket_client: AwsS3BucketClientImpl,
+}
+
+impl PublishedCollabS3StoreWithPostgresFallback {
+  pub fn new(
+    metrics: Arc<PublishedCollabMetrics>,
+    pg_pool: PgPool,
+    bucket_client: AwsS3BucketClientImpl,
+  ) -> Self {
+    Self {
+      metrics,
+      pg_pool,
+      bucket_client,
+    }
+  }
+}
+
+#[async_trait]
+impl PublishedCollabStore for PublishedCollabS3StoreWithPostgresFallback {
+  async fn publish_collabs(
+    &self,
+    publish_items: Vec<PublishCollabItem<serde_json::Value, Vec<u8>>>,
+    workspace_id: &Uuid,
+    user_uuid: &Uuid,
+  ) -> Result<(), AppError> {
+    let publish_items_batch_size = publish_items.len() as i64;
+    for publish_item in &publish_items {
+      check_collab_publish_name(publish_item.meta.publish_name.as_str())?;
+      let object_key = get_collab_s3_key(workspace_id, &publish_item.meta.view_id);
+      let data = publish_item.data.clone();
+      let bucket_client = self.bucket_client.clone();
+      let metrics = self.metrics.clone();
+      tokio::spawn(async move {
+        let result = bucket_client.put_blob(&object_key, &data).await;
+        if let Err(err) = result {
+          debug!("Failed to publish collab to S3: {}", err);
```

Comment:
maybe we can create a `mut vec![]` to store all the handles returned by `tokio::spawn`, then await them all. this will ensure that if the endpoint returns success, we have indeed put all the data in S3 and avoid situation where a success code is returned to client, but there's error putting in S3.

Reply:
Yeah, i was wondering whether we should do so. The downside of waiting for the request is, the time taken to make sure that the collab is written to both postgres and S3 will increase. Will need further discussion to determine whether the increase in latency is worth it or not, since we do have a fallback to postgres.